src/charm.py

```diff
@@ -86,6 +87,32 @@ class COSProxyCharm(CharmBase):
 
     _stored = StoredState()
 
+    mandatory_relation_pairs = {
```

Comment:
What is the intent of an item in `mandatory_relation_pairs`?  is it:
1. `{"name of mandatory relation": ["additional relations that we require one of", "if we have this relation"]}`
2. {"name of optional relation": ["list of relations that we require one of", "if we have this relation"]}`
?

based on [this](https://github.com/canonical/cos-proxy-operator/pull/145/files#diff-b9ed39bbc9c0387bd3e07da31d13373745534a1cd723d3e292c73496b12e307cR656) I think the intent is definition (1), but I don't think that solves #144.  #144 is a bug where cos-proxy is requiring the user to establish a `dashboards` relation even though they don't have/want one.  

I think here we're combining two concepts:
* a relation that the charm always requires (mandatory)
* a relation that the charm requires iff the charm has some other relation (eg: having one of `cos-agent` or `downstream-grafana-dashboard` if you have a `dashboards` relation)

where iiuc these should be separate.

Reply:
That `test_no_incoming_relations_blocks` passes is I think the problem.  If `test_no_incoming_relations_blocks` passes, then #144 will not be fixed

This test describes what #144 needs to pass:
```python
    def test_has_outgoing_dashboard_relation_without_incoming(self):
        """Assert charm is not Blocked if it has outgoing dashboard relation but not incoming."""
        self.harness.set_leader(True)

        # WHEN we have a downstream relation that can process dashboards
        downstream_rel_id = self.harness.add_relation(
            "downstream-grafana-dashboard", "cos-grafana"
        )
        self.harness.add_relation_unit(downstream_rel_id, "cos-grafana/0")

        # AND we have no dashboard relation

        # THEN the charm should not be blocked
        self.harness.evaluate_status()
        self.assertEqual(self.harness.model.unit.status, ActiveStatus())
```

I think the directionality of the mandatory pair is applied backwards here.  The test is saying "if I have a destination for dashboards, then I will block if I don't have a source".  But that's not a misconfigured case.  It would only be misconfigured if we have a source for dashboards and no target to process them

This feels most obvious if we have the `cos-agent` relation.  If we have a `cos-agent` target, then currently we require sources for all of `dashboards`, `monitors`, etc.  That sounds backwards.

Reply:
Yeah I see what you mean.  So the more correct test for #144 would be:

```python
    def test_has_outgoing_dashboard_relation_without_incoming(self):
        """Assert charm is not Blocked when we have a valid subset of the supported relations.
         
        This tests that if we have a cos-agent, which can consume a dashboard, but no incoming
        dashboards, that the charm does not block.
        """
        self.harness.set_leader(True)

        # WHEN we have a downstream relation that can process dashboards
        downstream_rel_id = self.harness.add_relation(
            "cos-agent", "cos-agent"
        )
        self.harness.add_relation_unit(downstream_rel_id, "cos-agent/0")

        # AND we have some other producer (a prometheus target)
        rel_id = self.harness.add_relation("prometheus-target", "target-app", app_data={
                "hostname": "scrape_target_0",
                "port": "1234",
            },)

        # AND we have no dashboard relation

        # THEN the charm should not be blocked
        self.harness.evaluate_status()
        self.assertEqual(self.harness.model.unit.status, ActiveStatus())
```

And you're right, it does pass for this PR.

Reply:
Test added in next commit.